### PR TITLE
Neopixel

### PR DESCRIPTION
**Pull request by yreda1999**
This is the ROS implementation of the Neopixels. The state of the neopixels can be set through a ROS service call. There is a callback aspect to the service. Currently, the service will wait indefinetly for an acknowledgement msg.

### CODE REVIEW

File: uwrt_mars_rover_neopixel/src/uwrt_mars_rover_neopixel_control.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      _can_interface(can_interface),
+      _neopixel_can_msg(NEOPIXEL_CAN_ID_OUTGOING, FRAME_PAYLOAD_LENGTH, can_interface) {
+  ros::init(argc, argv, "neopixel_set_server");
+}
+bool Neopixel::setState(uwrt_mars_rover_msgs::set_state::Request &req, uwrt_mars_rover_msgs::set_state::Response &res) {
+  ROS_INFO("Neopixel callback function triggered.");
+  _state = req.requested_mode.value;
+  _neopixel_can_msg.sendCAN(_state);
+  if (_neopixel_can_msg.waitforAck()) {
+    ROS_INFO("Received acknowledgement from gimbtonomy board.");
```
Comment (wmmc88) [inline uwrt_mars_rover_neopixel/src/uwrt_mars_rover_neopixel_control.cpp]: this should probably be a debug msg

File: uwrt_mars_rover_neopixel/src/uwrt_mars_rover_neopixel_control.cpp [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    // git CI insists on using 1u
+    res.success = 1u;
+    return res.success != 0u;
+  }
+  // As currently implemented, this will never run. See waitforAck() for more details
+  ROS_ERROR("Unexpected acknowledgement message received.");
+  res.success = 0u;
```
Comment (wmmc88) [inline uwrt_mars_rover_neopixel/src/uwrt_mars_rover_neopixel_control.cpp]: this shold be within the if statement above

File: uwrt_mars_rover_neopixel/src/uwrt_mars_rover_neopixel_control.cpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  return res.success != 0u;
+}
+void Neopixel::run() {
+  // Create handle for process
+  ros::NodeHandle neopixel_node;
```
Comment (wmmc88) [inline uwrt_mars_rover_neopixel/src/uwrt_mars_rover_neopixel_control.cpp]: this is a nodehandle technically, not a node

File: uwrt_mars_rover_neopixel/src/uwrt_mars_rover_neopixel_can.cpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+}
+void NeopixelCan::sendCAN(const uint8_t data) {
+  // Store data into the data potion of the CAN packet
+  _outgoing_packet.data[0] = data;
+  // Send out the CAN packet
+  write(_s, &_outgoing_packet, sizeof(_outgoing_packet));
+}
+bool NeopixelCan::waitforAck() {
+  ROS_INFO("Now waiting for acknowledgement message.");
+  bool first_pass = true;
```
Comment (wmmc88) [inline uwrt_mars_rover_neopixel/src/uwrt_mars_rover_neopixel_can.cpp]: no need for a first pass variable. should just have an if statement at the end that checks if the packet contents is valid

File: uwrt_mars_rover_neopixel/src/uwrt_mars_rover_neopixel_can.cpp [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  }
+}
+void NeopixelCan::sendCAN(const uint8_t data) {
+  // Store data into the data potion of the CAN packet
+  _outgoing_packet.data[0] = data;
+  // Send out the CAN packet
+  write(_s, &_outgoing_packet, sizeof(_outgoing_packet));
+}
+bool NeopixelCan::waitforAck() {
+  ROS_INFO("Now waiting for acknowledgement message.");
```
[the remark this answers is not in the document]
Comment (yreda1999) [reply uwrt_mars_rover_neopixel/src/uwrt_mars_rover_neopixel_can.cpp]: so u want all the ROS_INFO msgs to become ROS_INFO_NAMED("neopixel", <insert msg here>)?

[the remark this answers is not in the document]
Comment (wmmc88) [reply uwrt_mars_rover_neopixel/src/uwrt_mars_rover_neopixel_can.cpp]: more like `ROS_INFO_NAMED(name_, msg);` where `name_` is a member variable of a class that you set on initialization. Take a look at [this](https://github.com/uwrobotics/uwrt_mars_rover/blob/d29904f445a078f96ba8ae662d8449c2fcd79e63/uwrt_mars_rover_hw/include/uwrt_mars_rover_hw/uwrt_mars_rover_hw_drivetrain.h#L14) in our repo

File: uwrt_mars_rover_neopixel/src/uwrt_mars_rover_neopixel_control.cpp (hunk 2 above)
Comment (yreda1999) [reply uwrt_mars_rover_neopixel/src/uwrt_mars_rover_neopixel_control.cpp]: uhh why? the function returns true if it succeeds and we dont want to raise an error if it succeeds.

**Pull request merged**